prompt2model/model_trainer/qlora_trainer.py

```diff
@@ -0,0 +1,248 @@
+
+        if torch.cuda.device_count() > 1:  # If more than 1 GPU
+            self.model.is_parallelizable = True
+            self.model.model_parallel = True
+
+        output_dir = os.path.join(save_folder_path, "qlora")
+
+        self.eval_tokenizer = AutoTokenizer.from_pretrained(
+            self.model_name, add_bos_token=True, trust_remote_code=True
+        )
+
+        trainer = transformers.Trainer(
+            model=self.model,
+            train_dataset=train_dataset,
+            eval_dataset=eval_dataset,
+            args=transformers.TrainingArguments(
+                output_dir=output_dir,
+                num_train_epochs=num_epochs,
+                warmup_steps=5,
+                per_device_train_batch_size=train_batch_size,
+                gradient_checkpointing=True,
+                gradient_accumulation_steps=2,
+                weight_decay=0.001,
+                max_steps=-1,
+                learning_rate=lr,
+                logging_steps=100,
+                fp16=True,
+                optim="paged_adamw_8bit",
+                logging_dir="./logs",  # Directory for storing logs
+                save_strategy="steps",  # Save the model checkpoint every logging step
+                save_steps=200,  # Save checkpoints every 50 steps
+                evaluation_strategy="steps",  # Evaluate the model every logging step
+                eval_steps=100,  # Evaluate and save checkpoints every 50 steps
+                do_eval=True,  # Perform evaluation at the end of training
+                # report_to="wandb",  # Enable WandB logging
+                load_best_model_at_end=load_best_model_at_end,
+                metric_for_best_model="eval_loss",
+                greater_is_better=False,
+            ),
+            data_collator=transformers.DataCollatorForLanguageModeling(
+                self.tokenizer, mlm=False
+            ),
+            callbacks=[
+                EvalAccuracyCallback(original_eval_dataset, self.eval_tokenizer)
+            ],
+        )
+
+        self.model.config.use_cache = (
+            False  # silence the warnings. Please re-enable for inference!
+        )
+        trainer.train()
+
+        trainer.model.save_pretrained(os.path.join(output_dir, "qlora_model"))
+
+        del self.model
+        del trainer
+        del self.tokenizer
+        gc.collect()
+        torch.cuda.empty_cache()
+
+        self.model = AutoModelForCausalLM.from_pretrained(
+            self.model_name,  # Mistral, same as before
+            quantization_config=self.bnb_config,  # Same quantization config as before
+            device_map="auto",
+            trust_remote_code=True,
+        )
+
+        self.model = PeftModel.from_pretrained(
+            self.model, os.path.join(output_dir, "qlora_model")
+        )
+        self.model = self.model.merge_and_unload()
+        self.model.save_pretrained(os.path.join(output_dir, "final_model"))
```

Comment:
I'm wondering if the hard-coding of the model name here would cause problems, e.g. overwriting models. 

It makes sense to at least store this as a variable (e.g. `PEFT_MODEL_DIRECTORY=os.path.join(output_dir, "final_model")`) for better visibility

Reply:
the idea with passing `save_folder_path` to the function was to provide a path so that nothing in the current directory gets overwritten. but can also add the variable you mentioned for better visibility, and a comment in the demo